### PR TITLE
Fixes #1499 - Daily Value Raster Plot

### LINKED ISSUE #1499: Daily Value Raster Plot
**Background Information**: In [StreamStats](https://streamstats.usgs.gov/ss/), users can click on gages (represented as triangles on the map) to view more information about the gages, including a link to the "StreamStats Gage Page". When the user clicks the link to the StreamStats Gage Page, there will be 2 tabs-- one for Gage Information, and one for Gage Analysis Plots. On the Gage Analysis Plots tab, we will show a variety of plots, inspired by this page from the National Weather Service: https://www.weather.gov/source/aprfc/gageAnalysis.html?site=06893000. This issue will focus on adding a plot similar to the plot called "Daily Data" on that page. This plot is really two plots that are displayed next to one another. One is a plot of the daily flow value and the other is a plot of the annual average daily flow values. 

**Tasks**:  
- [x] Create daily plot.
    - [x] Use the `getDailyFlow` function to get all the daily data needed for this plot. Since both the Annual Flow Plot & this new Daily Data plot are using this same function, we should try to only call it once. This will limit the amount of service calls we make, thus making the application run more efficiently then if we call it separately for each plot. 
    - [x] Create a heatmap. Months should be on the x-axis, starting with Oct 1 (1st day of the water year) and ending with September 30th (last day of the water year), and water years on the y-axis. The values in each cell are based on the daily flow value for that date. It looks like the legend is on a logarithmic scale (see line 5383 of the national weather service code.)
    - [x] Add option to hover over cell and see a tool tip with the date, water year, and daily value.

---------------Feel free to break items below this line into new issue---------------

- [x] Create annual average plot.
    - [x] Create an annual heatmap. For this plot there should be just one column on the x axis, labeled Annual, and the y-axis should be in water years like the daily plot. To get the annual value, you need to take the average daily value of the entire water year. This will be the value used for each cell.
    - [x] Add option to hover over cell and see a tool tip with the water year and average daily value.
- [x] Display plots next to each other. In the national weather service code, it looks like the daily and annual are two different plots, see line 155 and 156, that are just displayed next to each other.  

**Tips**:
- If you go to https://www.weather.gov/source/aprfc/gageAnalysis.html?site=06893000, right click "View Source", you will be able to view the source code for that webpage along with all of its plots. The developer who wrote this page said we could look and use the code. The plot we want to recreate is called `heatchartobj` and `annualheatchartobj`.
    - They use heatmaps to create this plot, https://www.highcharts.com/demo/heatmap 
    - In function `setupGraphHeatQ`, you can see how the daily plot is formatted.  
    - In function `setupAnnualHeatQ`, you can see how the annual plot is formatted. 
- The graphing library we will be using is [HighCharts](https://www.highcharts.com/). 
- You will be working off of the branch called `gageplots`, primarily in the gagepage.html and GagePageController.ts. Pull requests should merge into the `gageplots` branch. We will need to manually close issues after merge requests are completed, since we are not merging into the default branch

### PR DESCRIPTION
**Pull request by maggiejaenicke**
**Logarithmic test sites:** 02173500, 1234000
**Linear test sites:** 02324500, 02231291, most places in Florida
Closes #1499.

### CODE REVIEW

File: src/Controllers/GagePageController.ts [hunk 1]
```diff
@@ -685,12 +695,88 @@ module StreamStats.Controllers {
...
                 this.dailyFlow.forEach(dailyObj => {
-                    if (dailyObj.qualifiers[0] === 'A') {
+                    if (parseInt(dailyObj.value) !== -999999) {
                     this.formattedDailyFlow.push({x: new Date(dailyObj.dateTime), y: parseInt(dailyObj.value)})
+                    this.dailyValuesOnly.push(parseInt(dailyObj.value));
                     }
                 });
             }
-            if (this.floodFreq) { //set up AEP plotLines
+            if (this.dailyFlow) {
```
Comment (harper-wavra) [inline src/Controllers/GagePageController.ts]: This if statement and the if statement at 696 are the same, so they could be combined into one. If we are keeping them apart for organization purposes, maybe we could add a comment by each if statement what (or what plot) it is being used for.

**Review by harper-wavra — CHANGES_REQUESTED** [review]
These plots are so pretty!

It does kind of freeze up for a second right after you select the 'Gage Analysis Plot' tab. I wonder if we could add some kind of loader. I think we have some loaders in the gagepage.html already so if we could get it to switch tabs right when you click that could solve it. If that isn't working, maybe we could add a full-page loader or perhaps a spinner next to the text on the 'Gage Analysis Plots' tab, just something to show the user that it is thinking. - feel free to break that into a different issue.

File: src/Controllers/GagePageController.ts [hunk 2]
```diff
@@ -883,6 +969,138 @@ module StreamStats.Controllers {
...
+                    text: 'Click and drag in the plot area to zoom in',
+                    align: 'center'
+                },
+                xAxis: {
+                    type: null,
+                    min: 275,
+                    max: 665,
+                    tickPositions: [275, 306, 336, 367, 398, 427, 458, 488, 519, 549, 580, 611, 650],
+                    title: {
+                        text: 'Date'
```
Comment (amedenblik) [inline src/Controllers/GagePageController.ts]:
```suggestion
                        text: 'Day of Year'
```

File: src/Controllers/GagePageController.js [hunk 3]
```diff
@@ -690,6 +762,146 @@ var StreamStats;
...
+                                        var month = fullDate.getUTCMonth();
+                                        month += 1;
+                                        var formattedUTCDate = month + '/' + UTCday + '/' + year;
+                                        var waterYear = year;
+                                        if (month > 9) {
+                                            waterYear += 1;
+                                        }
+                                        ;
+                                        if (doy > 282 && doy < 293)
+                                            return '</b><br>Water Year Average Value: <b>' + this.value.toFixed(2) + ' ft³/s</b><br>Water Year: <b>' + waterYear;
```
Comment (amedenblik) [inline src/Controllers/GagePageController.js]:
```suggestion
                                            return '</b><br>Water Year: <b>' + waterYear<br>Water Year Average Value: <b>' + this.value.toFixed(2) + ' ft³/s</b>;
```
Suggestion to put Water Year above the Water Year Average Value

File: src/Controllers/GagePageController.ts [hunk 4]
```diff
@@ -883,6 +969,138 @@ module StreamStats.Controllers {
...
+                        allowNegativeLog: true
+                    }
+                }
+            }
+            function isLeapYear(year) {
+                if (year % 400 === 0) return true;
+                if (year % 100 === 0) return false;
+                return year % 4 === 0;
+            }
+            this.heatChartConfig = {
```
Comment (amedenblik) [inline src/Controllers/GagePageController.ts:1022]: Is it possible to add a label to the legend that says "Flow (ft^3/s)"?

File: src/Views/gagepage.html [hunk 5]
```diff
@@ -224,13 +224,24 @@ <h4 class="modal-title">StreamStats Gage Page</h4>
...
+                    </center>
+                    <center>
+                        <div ng-if="!vm.heatChartConfig" class="panel panel-default reportSpinner">
+                            &nbsp;<i class="fa fa-spinner fa-3x fa-spin loadingSpinner"></i>
+                        </div>
+                        <div class="well well-sm" style="background-color:#EBF0F5;" ng-if="vm.heatChartConfig">
+                            <highchart id="chart2" config="vm.heatChartConfig"></highchart>
+                            <!-- <input type="checkbox" ng-if="vm.formattedDailyFlow.length > 0" id="logScale" name="logScale" ng-model="vm.logScale" ng-change="vm.logToLinear()" checked>
+                            <label for="logScale" ng-if="vm.formattedDailyFlow.length > 0"> Log Scale</label><br> -->
+                        </div>
```
Comment (amedenblik) [inline src/Views/gagepage.html]:
```suggestion

```

**Review by amedenblik — CHANGES_REQUESTED** [review]
Looks great! Nice work. I left a few comments.

File: src/Controllers/GagePageController.js (hunk 3 above)
Comment (amedenblik) [reply src/Controllers/GagePageController.js]: I am still seeing Water Year Average Value above Water Year
image

**Review by amedenblik — CHANGES_REQUESTED** [review]

File: src/Controllers/GagePageController.js (hunk 3 above)
Comment (maggiejaenicke) [reply src/Controllers/GagePageController.js]: Whoops I thought that one would automatically fix with the GitHub commit suggestion button. Just fixed it locally!

**Review by amedenblik — APPROVED** [review]

**Comment by amedenblik** [comment]
Update gageplots dev site: https://dev.streamstats.usgs.gov/gageplots/

**Pull request merged**